### DISCUSSION TITLE
[v2] Custom 404 error page: livewire component returns 404, when original 404 is caused by model not found and route model binding

### DISCUSSION THREAD

**Discussion by m-lotze** [livewire/livewire#7511] [7. Bugs (archived)] [answered]
### Livewire version

v2.12.6

### Laravel version

v10.30.0

### Which PHP version are you using?

PHP 8.3

### Steps To Reproduce

1. Add a livewire component to a custom 404 error page.
2. Add a route with implicit route model binding, e.g. model/{model}.
3. Load a 404 url, e.g. model/invalid-model-id
4. Make a request in the livewire component.
5. Livewire returns a 404 in a modal, instead of the actual response.

We use a custom 404 error page with various livewire components. This works as expected, if the 404 error occurs on a standard laravel route. However, when the route uses implicit model binding, any post request to livewire/message/... returns a 404.

### Repository or Wirebox URL

---

### Please confirm (incomplete submissions will not be addressed)

- [X] I have provided easy and step-by-step instructions to reproduce the bug.
- [X] I have provided code samples as text and NOT images.
- [X] I understand my bug report will be removed if I haven't met the criteria above.

**Comment by mgmbh** [comment]
Just a heads up, this bug is still present in v3.

Edit: I did some digging using Laravel Telescope and I noticed that on each /livewire/update request, the route model parameters are being resolved. It doesn't matter if you're on a custom error page or not, this happens on every request. But you don't notice it on the other ones because no 404 is thrown.
It's like the referer URL is being resolved, but I tried faking the Referer header via a browser extension and the result was still the same. I don't know enough about the internals of Livewire or Laravel to debug this further.

So in my view, the problem is that the update request is doing these queries in the first place. This leads to overhead since there is no reason to do these queries, or am I wrong?

**Reply by maltonite** [reply]
We have the same issue. Looking through the code, the start of the issue (in V2 at least) is In Livewire\Controllers\HttpConnectionHandler , within the applyPersistentMiddleware method. The correct 404 response is then thrown.

`   // Now run the faux request through the original middleware with a custom pipeline.
        (new Pipeline(app()))
            ->send($request)
            ->through($filteredMiddleware)
            ->then(function() {
                return new Response();
            });`

We are using a custom 404 page as well. 

Just returning a 404 status code within a controller method though doesn't cause this behaviour when used with the controller's view - we sometimes have to use 404 status codes, even if to the user the page is displaying correctly, all to do with how Google handles "google for jobs".

**Reply by maltonite** [reply]
In the end  I found that if I included the following in `routes/web.php`

`Route::fallback(function () {
    return response()->view('errors.404', [], 404);
})->name('fallback');`

Fixed the issue. This was due to the fact that in `Livewire\Controllers\HttpConnectionHandler` was looking for a route, which it would then mock for middleware applying purposes. The line in V2 where I think this happens is `$route = app('router')->getRoutes()->match($request);` The code wasn't finding a valid route, so would through a 404.

**Comment by m-lotze** [comment]
> Fixed the issue. This was due to the fact that in Livewire\Controllers\HttpConnectionHandler was looking for a route, which it would then mock for middleware applying purposes. The line in V2 where I think this happens is $route = app('router')->getRoutes()->match($request); The code wasn't finding a valid route, so would through a 404.

This does not solve the original issue. Livewire will find a valid route - the one which uses implicit model binding. It will not use the fallback route. And when it applies the middleware a `ModelNotFoundException` is thrown, which leads to a 404 error.

**Comment by philippbuerger** [comment]
Having the exact same issue. The solution mentioned above is not fixing it for me.

**Comment by nox-4O4** [comment, marked as answer]
I was able to circumvent the error by providing a `missing` callback with my route definition that gets used when the model binding could not be resolved. The `missing` callback returns the error page. 
This way no exceptions are thrown when the model binding fails, the Livewire component works, and the custom 404 page is still displayed.

```php
// Setting custom missing callback prevents SubstituteBindings from throwing an exception
// which breaks Livewire component updates on error pages.
Route::group(
    [
        'missing' => fn() => response()->view('errors.404')->setStatusCode(404),
    ],
    function () {
        // some routes with model bindings
        Route::get('/users/{user}', UserDetail::class)->name('users.display');
    }
);
```

If you don't want to group the routes, you can also call `->missing(...)` when adding the route:
```php
Route::get(/* something with model bindings */)->missing(fn() => response()->view('errors.404')->setStatusCode(404));
```
